### DISCUSSION TITLE
Type narrowing while comparing to "type()" of a class instance

### DISCUSSION THREAD

**Discussion by Delgan** [microsoft/pyright#9520] [Ideas]
Hi.

I've observed that type narrowing doesn't work in a situation where the type of an unknown variable is compared with the type of a known class instance.

For example (pyright 1.1.389):

```python
from typing import Any, reveal_type

class A:
    pass

a = A()

reveal_type(type(a))  # Type of "type(a)" is "Type[A]"

def func_1(var: A | None):
    if type(var) is type(a):
        reveal_type(var)  # Type of "A | None"


def func_2(var: Any):
    if type(var) is type(a):
        reveal_type(var)  # Type of "var" is "A"
```

Curiously, it works as expected when `var` is `Any` but not when var is `A | None`. I also tested with `isinstance()` and `a.__class__` without success.

Is this a pattern you might consider supporting?

My specific use case involves situations where class `A` is not publicly accessible from a third-party library, even though instances of the class are.

**Comment by erictraut** [comment]
This is not a supported type guard form. For a full list of supported type guards, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

We do occasionally add support for new forms, but each requires custom logic and can potentially have a measurable impact on code analysis performance, so the bar is high. I wouldn't consider supporting this pattern at this time because this is the first request I've ever received for it. I'm willing to reconsider in the future if there's a signal from enough other pyright users (e.g. in the form of upvotes on this discussion topic).
